src/event.ts

```diff
@@ -77,47 +70,182 @@ export function toEventHandler (handler: CompatibilityEventHandler): EventHandle
-  // @ts-ignore
-  event.event = event
-  // @ts-ignore
-  req.event = event
-  // @ts-ignore
-  req.context = event.context
-  // @ts-ignore
-  req.req = req
-  // @ts-ignore
-  req.res = res
-  // @ts-ignore
-  res.event = event
-  // @ts-ignore
-  res.res = res
-  // @ts-ignore
-  res.req = res.req || {}
-  // @ts-ignore
-  res.req.res = res
-  // @ts-ignore
-  res.req.req = req
-
-  return event
+export function createEvent (req: http.IncomingMessage, res: http.ServerResponse): H3Event {
+  return new H3Event(req, res)
+}
+
+export class H3Event {
+  '__is_event__' = true
+  req: IncomingMessage
+  res: ServerResponse
+  event: H3Event
+  context: Record<string, any> = {}
+
+  constructor (req: http.IncomingMessage | IncomingMessage, res: http.ServerResponse | ServerResponse) {
+    this.req = req as IncomingMessage
+    this.res = res as ServerResponse
+
+    // Backward comatibility for interchangable usage of {event,req,res}.{req,res}
+    // TODO: Remove in future versions
+    this.event = this
+    // @ts-ignore
+    req.event = this
+    // @ts-ignore
+    req.context = this.context
+    // @ts-ignore
+    req.req = req
+    // @ts-ignore
+    req.res = res
+    // @ts-ignore
+    res.event = this
+    // @ts-ignore
+    res.res = res
+    // @ts-ignore
+    res.req = res.req || {}
+    // @ts-ignore
+    res.req.res = res
+    // @ts-ignore
+    res.req.req = req
+  }
+
+  // Implementation of FetchEvent
+  respondWith <T> (r: HandlerResponse<T> | H3Response | Promise<H3Response>): void {
+    Promise.resolve(r).then((_response) => {
+      if (this.res.writableEnded) { return }
+
+      const response = _response instanceof H3Response ? _response : new H3Response(_response)
+
+      response.headers.forEach((value, key) => {
+        this.res.setHeader(key, value)
+      })
+      if (response.status) {
+        this.res.statusCode = response.status
+      }
+      if (response.statusText) {
+        this.res.statusMessage = response.statusText
+      }
+      if (response.redirected) {
+        this.res.setHeader('Location', response.url)
+        return this.res.end()
+      }
+      if (!response._body) {
+        return this.res.end()
+      }
+      if (typeof response._body === 'string' || 'buffer' in response._body || 'byteLength' in response._body) {
+        return this.res.end(response._body)
+      }
+      if (!response.headers.has('content-type')) {
+        response.headers.set('content-type', MIMES.json)
+      }
+      this.res.end(JSON.stringify(response._body, null, 2))
+    })
+  }
 }
 
 export function isEvent (input: any): input is H3Event {
   return '__is_event__' in input
 }
+
+export class H3Headers implements Headers {
+  _headers: Record<string, string>
+
+  constructor (init?: HeadersInit) {
+    if (!init) {
+      this._headers = {}
+    } else if (Array.isArray(init)) {
+      this._headers = Object.fromEntries(init.map(([key, value]) => [key.toLowerCase(), value]))
```

Comment:
```suggestion
      this._headers = init.reduce((acc, [key, value]) => {
        const lk = key.toLowerCase()
        acc[lk] = acc[lk] != null ? `${acc[lk]}, ${value}` : value
        return acc
      }, {})
```

I think it is better to use `reduce` because duplicate keys may be specified.

```ts
const headers = new Headers([['Example', 'foo'], ['Example', 'bar']])
// headers.get('Example') === 'foo, bar'
```